### PR TITLE
Adding a local executable plugin mechanism.

### PR DESCRIPTION
**Pull request by benvanik**
The plugin API is defined in the standalone executable_plugin.h header.
This is similar to executable_library.h and enables building compatible
plugins without relying on any other IREE code. For plugins that do,
though, support for iree_status_t integration for propagating useful
errors across the plugin boundry is supported.

A plugin manager provides lifetime management and import provider
registration such that we can have stateful import providers (JITs,
dynamically loaded libraries, etc). To keep the import feature optional
the manager sits behind the import provider interface and will only be
pulled in when the user actually creates the manager. There's definitely
room for improvement here on naming and simplification but this should
be enough to start experimenting with plugins.  Plugins are scanned in
reverse order for imports allowing for overrides of
builtin/statically-registered plugins by dynamic ones.

To start we have 3 plugin types built in: static, system-library, and
embedded-elf. Similar to the static executable loader the static plugin
allows for plugins statically linked into the application binary to be
registered for use. The other two allow for dynamic loading of plugins
at runtime.

The embedded ELF loader can be used when compiling in a standalone mode
to enable portable/hermetic plugin deployments. For ukernel-like plugins
that export specialized kernels this gives the same cross-platform
portability that our embedded ELF executables have. FatELF support has
been added to allow for portable plugins when using the standalone ELFs.
More complex software (JITs, etc) can use the system loader and do
sketchy stuff with ease (TLS, global state, threading, etc).

For easy integration in specialized runtime builds the
`IREE_HAL_EXECUTABLE_PLUGIN_REGISTRATION_FN` define is available to
hook into the initial plugin manager registration and give the hosting
application the ability to inject its own plugins. Otherwise the dynamic
registration will work either via the `--executable_plugin=` repeated
flag or explicit registration for users not going through the
driver_module registration path.

All IREE tools/bindings using the driver_module mechanism will get the
new flag and can specify plugins as file paths. If the `embedded:`
prefix is used or the file extension is `.sos` the embedded loader will
be used and otherwise the system one will be. Both of those types can
be toggled at a cmake level based on availability (bare metal won't
have the system loader, etc).

Examples:
```
--executable_plugin=some/system.dll
--executable_plugin=some/standalone.sos
--executable_plugin=embedded:some/standalone.so
```

NOTE: MSVC support for standalone executable plugins is broken
until #12801 and the sample has been disabled there.

### CODE REVIEW

File: runtime/src/iree/hal/local/elf/arch/x86_64_msvc.asm [hunk 1]
```diff
@@ -185,19 +185,59 @@ iree_elf_call_i_ppp PROC FRAME
...
+
+; DO NOT SUBMIT
+; going elf->elf needs to shift out pointer
+; could change ABI to avoid this (rdi/rsi/rdx are args, rdi is ptr)
+;
+; only one thunk per table, prevents mixed plugins
+;
+; int iree_elf_thunk_i_ppp(const void* symbol_ptr, void* a0, void* a1, void* a2)
+iree_elf_thunk_i_ppp PROC
+  ; _sysv_interop_prolog
```
Comment (ScottTodd) [inline runtime/src/iree/hal/local/elf/arch/x86_64_msvc.asm]: reminder to resolve this somehow and remove the `DO NOT SUBMIT` before merge

File: runtime/src/iree/hal/local/executable_loader.h [hunk 2]
```diff
@@ -47,24 +58,35 @@ iree_hal_executable_import_provider_null(void) {
...
+// Returns true if the import |symbol_name| is optional.
+static bool iree_hal_executable_import_is_optional(const char* symbol_name) {
+  // A `?` prefix indicates the symbol is optional and can be NULL.
+  // Since the strings are NUL terminated we know there's always 1 char and
+  // we can just test that for the prefix.
+  return symbol_name ? (symbol_name[0] == '?') : false;
+}
 
 //===----------------------------------------------------------------------===//
 // iree_hal_executable_loader_t
```
Comment (ScottTodd) [inline runtime/src/iree/hal/local/executable_loader.h:92]: We could use some system architecture diagrams for executables and loaders in the local HAL README (https://github.com/openxla/iree/tree/main/runtime/src/iree/hal/local), especially if we're expecting hosting applications or downstream projects to add their own plugins/loaders/etc.

(and diagrams on https://openxla.github.io/iree/extensions/ would be neat too)

File: runtime/src/iree/hal/local/executable_plugin.h [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+// Versioning and interface querying
+//===----------------------------------------------------------------------===//
+
+// Version code indicating the minimum required runtime structures.
+// Runtimes cannot load plugins with newer versions but may be able to load
+// older versions if backward compatibility is enabled.
+//
+// NOTE: until we hit v1 the versioning scheme here is not set in stone.
+// We may want to make this major release number, date codes (0x20220307),
+// or some semantic versioning we track in whatever spec we end up having.
```
Comment (ScottTodd) [inline runtime/src/iree/hal/local/executable_plugin.h:133]: Did you intend for this example date code to be the date you wrote the code? 2023 not 2022 in that case :P

```suggestion
// NOTE: until we hit v1 the versioning scheme here is not set in stone.
// We may want to make this major release number, date codes (0x20230307),
// or some semantic versioning we track in whatever spec we end up having.
```

File: runtime/src/iree/hal/local/executable_plugin.h [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+    void* self, const iree_hal_executable_plugin_resolve_params_v0_t* params,
+    iree_hal_executable_plugin_resolution_t* out_resolution);
+
+// Structure used for v0 plugin interfaces.
+// The entire structure is designed to be read-only and able to live embedded in
+// the binary .rdata section.
+//
+// Thread-safe: the plugin must be safe to load and resolve from multiple
+// threads simultaneously. No unloads will be performed with active resolves.
+typedef struct iree_hal_executable_plugin_v0_t {
```
Comment (ScottTodd) [inline runtime/src/iree/hal/local/executable_plugin.h:397]: Can you explain here (or at the top of the file) what an executable plugin is / what they can do? (I'm still a bit vague on the details even after reading the PR description and this code)

File: samples/custom_dispatch/cpu/plugin/README.md
Comment (ScottTodd) [inline samples/custom_dispatch/cpu/plugin/README.md:1]: Empty README?

File: samples/custom_dispatch/cpu/plugin/standalone_plugin.c [hunk 5]
```diff
@@ -0,0 +1,165 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// Demonstrates a standalone plugin exporting a single `import_mul` function.
+// This models kernel libraries and custom intrinsics where there's only
+// stateless functions. Standalone plugins can be compiled to portable ELFs that
```
Comment (ScottTodd) [inline samples/custom_dispatch/cpu/plugin/standalone_plugin.c]: nit: grammar (also, this sample helps me understand how plugins fit into the broader architecture, thanks!)

```suggestion
// Demonstrates a standalone plugin exporting a single `import_mul` function.
// This models kernel libraries and custom intrinsics where there are only
// stateless functions. Standalone plugins can be compiled to portable ELFs that
```

File: samples/custom_dispatch/cpu/plugin/standalone_plugin.c [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    const float* restrict binding1_aligned;
+    size_t binding1_offset;
+    size_t binding1_size;
+    size_t binding1_stride;
+    float* restrict binding2;
+    float* restrict binding2_aligned;
+    size_t binding2_offset;
+    size_t binding2_size;
+    size_t binding2_stride;
+    // ^^^^ all this is garbage and should just be 3 pointers, but MLIR
```
Comment (ScottTodd) [inline samples/custom_dispatch/cpu/plugin/standalone_plugin.c]: Ouch :/

File: samples/custom_dispatch/cpu/plugin/example.mlir [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+      // values passed in the dispatch operation (usually things like root
+      // output op tensor dimensions and other abstract values).
+      %x = affine.apply affine_map<()[s0] -> (s0 ceildiv 64)>()[%workload]
+      %c1 = arith.constant 1 : index
+      stream.return %x, %c1, %c1 : index, index, index
+    }
+
+    builtin.module {
+      // External function declaration using a user-chosen calling convention.
+      func.func private @simple_mul_workgroup(%binding0: memref<?xf32>, %binding1: memref<?xf32>, %binding2: memref<?xf32>, %dim: index, %tid: index)
```
Comment (ScottTodd) [inline samples/custom_dispatch/cpu/plugin/example.mlir]: Should this have an attribute or use something other than `func.func`? I'd expect external / plugin-defined functions to be easier to spot (control+F to search for across a module, etc.)

File: runtime/src/iree/hal/local/executable_plugin.h (hunk 3 above)
Comment (benvanik) [reply runtime/src/iree/hal/local/executable_plugin.h:133]: it's just an example :)
image

File: samples/custom_dispatch/cpu/plugin/example.mlir (hunk 7 above)
Comment (benvanik) [reply samples/custom_dispatch/cpu/plugin/example.mlir]: externally defined functions are just externally defined functions - nothing special about them - lets us use upstream lowerings to func.call and such without requiring IREE-specific internals to leak (especially useful in plugins that can use just upstream dialects)

File: samples/custom_dispatch/cpu/plugin/README.md
Comment (benvanik) [reply samples/custom_dispatch/cpu/plugin/README.md:1]: I've been procrastinating on this... ugh :P

File: samples/custom_dispatch/cpu/plugin/example.mlir [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+        // Bindings are accessed by reference.
+        %memref0 = stream.binding.subspan %binding0[%c0] : !stream.binding -> memref<?xf32>{%dim}
+        %memref1 = stream.binding.subspan %binding1[%c0] : !stream.binding -> memref<?xf32>{%dim}
+        %memref2 = stream.binding.subspan %binding2[%c0] : !stream.binding -> memref<?xf32>{%dim}
+
+        // Call the externally defined C function with an (almost) plain C
+        // calling convention (see above for details about the mess memrefs
+        // turn into). This will be fetched at runtime from the plugin binary.
+        func.call @simple_mul_workgroup(%memref0, %memref1, %memref2, %dim, %tid) : (memref<?xf32>, memref<?xf32>, memref<?xf32>, index, index) -> ()
```
Comment (ScottTodd) [inline samples/custom_dispatch/cpu/plugin/example.mlir]: LGTM once this test passes

Logs: https://github.com/openxla/iree/actions/runs/4696175348/jobs/8326087188?pr=12625#step:6:2625
```
/work/samples/custom_dispatch/cpu/plugin/example.mlir:117:9: error: 'llvm.load' op result #0 must be LLVM type with size, but got '!llvm.func<i32 (ptr<i8>, ptr<i8>, ptr<i8>)>'
        func.call @simple_mul_workgroup(%memref0, %memref1, %memref2, %dim, %tid) : (memref<?xf32>, memref<?xf32>, memref<?xf32>, index, index) -> ()
        ^
/work/samples/custom_dispatch/cpu/plugin/example.mlir:117:9: note: see current operation: %52 = "llvm.load"(%48) {ordering = 0 : i64} : (!llvm.ptr<ptr>) -> !llvm.func<i32 (ptr<i8>, ptr<i8>, ptr<i8>)>
```

**Review by ScottTodd — APPROVED** [review]

File: samples/custom_dispatch/cpu/plugin/example.mlir (hunk 8 above)
Comment (benvanik) [reply samples/custom_dispatch/cpu/plugin/example.mlir]: darn looks like a merge issue - will fix!

**Comment by lundong** [comment]
@benvanik  Hi Ben, it looks that static plugins are not fully supported, unlike system-library and
embedded-elf. Could you please help add the full support (with samples if possible) for static plugins? We expect to be using it on ukernel plugins. Thanks a lot.
@hcindyl

**Comment by benvanik** [comment]
Static plugins are supported - you can either use the full plugin interface and `iree_hal_executable_plugin_manager_register_plugin` to register your plugin or the lighter-weight `iree_hal_executable_plugin_manager_register_provider` that just provides a function name -> function pointer lookup.

**Comment by lundong** [comment]
> Static plugins are supported - you can either use the full plugin interface and `iree_hal_executable_plugin_manager_register_plugin` to register your plugin or the lighter-weight `iree_hal_executable_plugin_manager_register_provider` that just provides a function name -> function pointer lookup.

@benvanik Unlike system_library_loader and embedded_elf_loader, it looks to me that static_library_loader doesn't have visibility on plugin_manager: executable_plugin_manager.h is not included in static_library_loader.c, and iree_hal_static_library_loader_create() doesn't have the plugin_manager argument. Can you please further clarify on it? Thank you very much.

**Comment by benvanik** [comment]
`iree_hal_static_library_loader_create` takes an `iree_hal_executable_import_provider_t` - you can get the provider from the plugin manager with `iree_hal_executable_plugin_manager_provider`.

Note that if you are using static library executable compilation you will want to avoid using imports if possible - the benefit of statically linking the codegen output _and_ your custom functions is that they can be made as efficient direct function calls and during compilation of your binary the linker can drop unused functions, inline functions, etc. Using static library executables and statically linked import providers/plugins is useful only for optionality (same code may be deployed statically and dynamically, etc) but otherwise it's not something that should be done unless you don't care about code size or execution efficiency.

**Pull request merged**